### PR TITLE
 Change run selection in 'simex e', add documentation on CLI, and minor fixes

### PR DESCRIPTION
**Pull request by duylethanh**
The PR contains the following:
- ~``args.run`` will now be compared to ``run.experiment.display_name + '/' + run.instance.shortname`` instead of ``run.experiment.name + '/' + run.instance.shortname``~
- ``--run`` will now be parsed into experiment, variants, revision, instance and repetition
- ``--repetition``was added to the ``run_selection_parser``
- logging messages when launching/skipping/purging a run are now more detailed
- ``Config.check_for_reserved_name()`` will also check for characters in ``[' ', '~', '@', '/', '[', ']']`` and raise a ``RuntimeError`` if they are detected
- updated documentation of 'experiments'  and added sections on 'builds' and 'develop' in command-line reference
- minor fixes (typos, redundant parentheses, rename variable)

### CODE REVIEW

**Comment by avdgrinten** [comment]
A better strategy for `--run` would consist of parsing the argument into revision, variants, repetition, etc. That way, we could implement the selection in a faster way and we would be independent of whitespace and other syntactic differences.

**Comment by duylethanh** [comment]
PR and PR description above updated. For the ``--run`` parsing I used the following regex:

```
r'(?P<experiment>[^~@/\[\]]+)'
r'(~(?P<variants>[^~@/\[\]]+))?'
r'(@(?P<revision>[^~@/\[\]]+))?'
r'(/(?P<instance>[^~@/\[\]]+))'
r'(\[(?P<repetition>\d+)\])?'
```

It parses the format ``<experiment_name>~<variants>@<revision_name>/<instance_name>[<repetition>]``, where ``~<variants>``, ``~<revision>`` and ``~[<repetition>]`` are optional and ``<variants>`` should be a comma separated list of variant names. ``?P<name>``allows to name groups so that they can be addressed more easily. ``[^~@/\[\]]`` allows any character except the delimiters we are using to separate variants, revisions, instances and repetitions. If ``<[repetition]>`` is not provided, it will default to ``0``.

**Comment by avdgrinten** [comment]
Alright, that looks fine. I will need to read the docs though.

**Comment by duylethanh** [comment]
``Config.check_for_reserved_name()`` will also check for characters in ``[' ', '~', '@', '/', '[', ']']`` and raise a ``RuntimeError`` if they are detected

**Comment by duylethanh** [comment]
I only removed the merge conflicts.

File: docs/command_line_reference.rst [hunk 1]
```diff
@@ -10,47 +10,97 @@ Simexpal instructions are written as:
...
+:``delete-source``: deletes the source directory when purging.
+:``install``: (re-)installs the build.
+:``purge``: deletes all related build files.
+:``recheckout``: deletes the local build Git repository, reclones, regenerates, reconfigures, recompiles
+    and reinstalls it.
+:``recompile``: recompiles and reinstall the build.
+:``reconfigure``: reconfigures, recompiles and reinstalls the build.
+:``regenerate``: (re-)regenerates the build.
+:``reinstall``: reinstalls the build.
+:``reregenerate``: regenerates, reconfigures, recompiles and reinstalls the build.
```
Comment (avdgrinten) [inline docs/command_line_reference.rst:55]: Better specify these as `--checkout` etc.

Comment (avdgrinten) [reply docs/command_line_reference.rst:55]: The same applies to other commands.

File: scripts/simex [hunk 2]
```diff
@@ -67,8 +67,54 @@ def cli_selects_run(args, run):
...
+		if args.repetition != run.repetition:
+			return False
 	if args.run is not None:
-		if args.run != run.experiment.name + '/' + run.instance.shortname:
+
+		m = re.search(r'(?P<experiment>[^~@/\[\]]+)'
+						r'(~(?P<variants>[^~@/\[\]]+))?'
+						r'(@(?P<revision>[^~@/\[\]]+))?'
+						r'(/(?P<instance>[^~@/\[\]]+))'
+						r'(\[(?P<repetition>\d+)\])?', args.run)
```
Comment (avdgrinten) [inline scripts/simex:79]: We should compile this regexp outside of the cli_selects_run function to avoid redundant work.

**Review by avdgrinten — COMMENTED** [review]
Some final comments, remainder looks good.

File: docs/command_line_reference.rst (hunk 1 above)
Comment (duylethanh) [reply docs/command_line_reference.rst]: Done.

File: scripts/simex (hunk 2 above)
Comment (duylethanh) [reply scripts/simex]: Done.

**Comment by duylethanh** [comment]
PR updated.

**Comment by avdgrinten** [comment]
LGTM now.

**Pull request merged**